File: web/src/commands/lists.ts
import type { CrosstuneDb } from '../db/schema'
import type { LocalListItem } from '../db/types'
import { moveBeside } from '../features/lists/order'
import { activeByPosition, newId, nextPosition, now, putRow, tombstone, writeTx } from './write'

export async function activeItems(db: CrosstuneDb, listId: string): Promise<LocalListItem[]> {
  return activeByPosition(await db.list_items.where('list_id').equals(listId).toArray())
}

export async function createList(db: CrosstuneDb, name: string): Promise<string> {
  const trimmed = name.trim()
  if (!trimmed) throw new Error('A list needs a name')
  const at = now()
  const id = newId()
  await writeTx(db, async () => {
    const lists = await db.lists.toArray()
    const position = nextPosition(lists.filter((list) => !list.deleted_at))
    await putRow(db, 'lists', {
      id,
      created_at: at,
      updated_at: at,
      deleted_at: null,
      server_seq: 0,
      name: trimmed,
      position,
    })
  })
  return id
}

export async function renameList(db: CrosstuneDb, listId: string, name: string): Promise<void> {
  const trimmed = name.trim()
  if (!trimmed) throw new Error('A list needs a name')
  await writeTx(db, async () => {
    const list = await db.lists.get(listId)
    if (!list || list.deleted_at) throw new Error('List not found')
    await putRow(db, 'lists', { ...list, name: trimmed, updated_at: now() })
  })
}

export async function deleteList(db: CrosstuneDb, listId: string): Promise<void> {
  const at = now()
  await writeTx(db, async () => {
    await tombstone(db, 'lists', listId, at)
    const items = await db.list_items.where('list_id').equals(listId).toArray()
    for (const item of items) {
      await tombstone(db, 'list_items', item.id, at, { enqueueDelete: false })
    }
  })
}

export async function addToList(
  db: CrosstuneDb,
  listId: string,
  userSongId: string,
): Promise<string> {
  const at = now()
  const id = newId()
  return writeTx(db, async () => {
    const list = await db.lists.get(listId)
    if (!list || list.deleted_at) throw new Error('List not found')
    const items = await activeItems(db, listId)
    const existing = items.find((item) => item.user_song_id === userSongId)
    if (existing) return existing.id
    await putRow(db, 'list_items', {
      id,
      created_at: at,
      updated_at: at,
      deleted_at: null,
      server_seq: 0,
      list_id: listId,
      user_song_id: userSongId,
      position: nextPosition(items),
    })
    return id
  })
}

export async function removeFromList(db: CrosstuneDb, itemId: string): Promise<void> {
  await writeTx(db, () => tombstone(db, 'list_items', itemId, now()))
}

/** Renumber from 0, touching only the rows whose position changed. */
export async function writeOrder(db: CrosstuneDb, ordered: LocalListItem[]): Promise<void> {
  const at = now()
  for (const [position, item] of ordered.entries()) {
    if (item.position !== position) {
      await putRow(db, 'list_items', { ...item, position, updated_at: at })
    }
  }
}

/** Move an item just past the target; see moveBeside. */
export async function moveItem(
  db: CrosstuneDb,
  listId: string,
  itemId: string,
  targetItemId: string,
): Promise<void> {
  await writeTx(db, async () => {
    const items = await activeItems(db, listId)
    await writeOrder(
      db,
      moveBeside(items, (item) => item.id, itemId, targetItemId),
    )
  })
}

File: apple/CrosstuneKit/Sources/CrosstuneCommands/Lists.swift
import CrosstuneStore
import Foundation
import GRDB

extension StoreWriter {
    /// A list's items that have not been tombstoned, in position order.
    func activeItems(listID: String) throws -> [ListItem] {
        activeByPosition(try ListItem.filter(Column("list_id") == listID).fetchAll(db))
    }

    /// Creates a list past every other active list.
    @discardableResult
    public func createList(_ name: String, at time: Timestamp = .now) throws -> String {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { throw CommandError.listNameRequired }
        let active = try TuneList.fetchAll(db).filter { $0.deletedAt == nil }
        let id = newID(at: time)
        try put(TuneList(id: id, createdAt: time, name: trimmed, position: nextPosition(active)), at: time)
        return id
    }

    /// Renames a live list. Refuses a name that trims to nothing or a missing list.
    public func renameList(_ listID: String, name: String, at time: Timestamp = .now) throws {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { throw CommandError.listNameRequired }
        guard var list = try TuneList.fetchOne(db, key: listID), list.deletedAt == nil else {
            throw CommandError.listNotFound
        }
        list.name = trimmed
        try put(list, at: time)
    }

    /// Tombstones a list and its items, queuing only the list's own delete.
    public func deleteList(_ listID: String, at time: Timestamp = .now) throws {
        try tombstone(TuneList.self, id: listID, at: time)
        let items = try ListItem.filter(Column("list_id") == listID).fetchAll(db)
        for item in items {
            try tombstone(ListItem.self, id: item.id, at: time, enqueueDelete: false)
        }
    }

    /// Adds a tune to a list past its other items, or returns the item already there.
    @discardableResult
    public func addToList(_ listID: String, userTuneID: String, at time: Timestamp = .now) throws -> String {
        guard let list = try TuneList.fetchOne(db, key: listID), list.deletedAt == nil else {
            throw CommandError.listNotFound
        }
        let items = try activeItems(listID: listID)
        if let existing = items.first(where: { $0.userTuneID == userTuneID }) {
            return existing.id
        }
        let id = newID(at: time)
        try put(
            ListItem(id: id, createdAt: time, listID: listID, userTuneID: userTuneID, position: nextPosition(items)),
            at: time)
        return id
    }

    public func removeFromList(_ itemID: String, at time: Timestamp = .now) throws {
        try tombstone(ListItem.self, id: itemID, at: time)
    }

    /// Renumbers `ordered` from 0, writing only the items whose position changed.
    public func writeOrder(_ ordered: [ListItem], at time: Timestamp = .now) throws {
        for (position, item) in ordered.enumerated() where item.position != position {
            var moved = item
            moved.position = position
            try put(moved, at: time)
        }
    }

    /// Moves an item just past the target; see ``moveBeside(_:id:itemID:targetID:)``.
    public func moveItem(listID: String, itemID: String, targetID: String, at time: Timestamp = .now) throws {
        let items = try activeItems(listID: listID)
        try writeOrder(moveBeside(items, id: \.id, itemID: itemID, targetID: targetID), at: time)
    }
}

extension Commands {
    @discardableResult
    public func createList(_ name: String, at time: Timestamp = .now) async throws -> String {
        try await store.write { writer in try writer.createList(name, at: time) }
    }

    public func renameList(_ listID: String, name: String, at time: Timestamp = .now) async throws {
        try await store.write { writer in try writer.renameList(listID, name: name, at: time) }
    }

    public func deleteList(_ listID: String, at time: Timestamp = .now) async throws {
        try await store.write { writer in try writer.deleteList(listID, at: time) }
    }

    @discardableResult
    public func addToList(_ listID: String, userTuneID: String, at time: Timestamp = .now) async throws -> String {
        try await store.write { writer in try writer.addToList(listID, userTuneID: userTuneID, at: time) }
    }

    public func removeFromList(_ itemID: String, at time: Timestamp = .now) async throws {
        try await store.write { writer in try writer.removeFromList(itemID, at: time) }
    }

    public func moveItem(listID: String, itemID: String, targetID: String, at time: Timestamp = .now) async throws {
        try await store.write { writer in
            try writer.moveItem(listID: listID, itemID: itemID, targetID: targetID, at: time)
        }
    }
}
